component/stream/read_write_filename.go

```diff
@@ -0,0 +1,478 @@
+   copies or substantial portions of the Software.
+
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+   SOFTWARE
+*/
+
+package stream
+
+import (
+	"encoding/base64"
+	"errors"
+	"io"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/Azure/azure-storage-fuse/v2/common"
+	"github.com/Azure/azure-storage-fuse/v2/common/log"
+	"github.com/Azure/azure-storage-fuse/v2/internal"
+	"github.com/Azure/azure-storage-fuse/v2/internal/handlemap"
+	"github.com/pbnjay/memory"
+)
+
+type ReadWriteFilenameCache struct {
+	sync.RWMutex
+	*Stream
+	StreamConnection
+	fileCache map[string]*handlemap.Cache
+}
+
+func (rw *ReadWriteFilenameCache) Configure(conf StreamOptions) error {
+	if conf.BufferSize <= 0 || conf.BlockSize <= 0 || conf.CachedObjLimit <= 0 {
+		rw.StreamOnly = true
+	}
+	rw.BlockSize = int64(conf.BlockSize) * mb
+	rw.BufferSize = conf.BufferSize * mb
+	rw.CachedObjLimit = int32(conf.CachedObjLimit)
+	rw.fileCache = make(map[string]*handlemap.Cache)
+	rw.CachedObjects = 0
+	return nil
+}
+
+func (rw *ReadWriteFilenameCache) CreateFile(options internal.CreateFileOptions) (*handlemap.Handle, error) {
+	// log.Trace("Stream::CreateFile : name=%s, mode=%s", options.Name, options.Mode)
+	handle, err := rw.NextComponent().CreateFile(options)
+	if err != nil {
+		log.Err("Stream::CreateFile : error failed to create file %s: [%s]", options.Name, err.Error())
+	}
+	if !rw.StreamOnly {
+		err = rw.createFileCache(handle)
+		if err != nil {
+			log.Err("Stream::CreateFile : error creating cache object %s [%s]", options.Name, err.Error())
+		}
+	}
+	return handle, err
+}
+
+func (rw *ReadWriteFilenameCache) OpenFile(options internal.OpenFileOptions) (*handlemap.Handle, error) {
+	// log.Trace("Stream::OpenFile : name=%s, flags=%d, mode=%s", options.Name, options.Flags, options.Mode)
+	handle, err := rw.NextComponent().OpenFile(options)
+	if err != nil {
+		log.Err("Stream::OpenFile : error failed to open file %s [%s]", options.Name, err.Error())
+		return handle, err
+	}
+	if !rw.StreamOnly {
+		err = rw.createFileCache(handle)
+		if err != nil {
+			log.Err("Stream::OpenFile : error failed to create cache object %s [%s]", options.Name, err.Error())
+		}
+	}
+	return handle, err
+}
+
+func (rw *ReadWriteFilenameCache) ReadInBuffer(options internal.ReadInBufferOptions) (int, error) {
+	// log.Trace("Stream::ReadInBuffer : name=%s, handle=%d, offset=%d", options.Handle.Path, options.Handle.ID, options.Offset)
+	if !rw.StreamOnly && options.Handle.CacheObj.StreamOnly {
+		err := rw.createFileCache(options.Handle)
```

Comment:
If handle is in streaming only mode, why do we need to create the fileCache. It shall just get required data from upstream component and return it back.

Reply:
The reason is if the handle is read only but the mount is not read only then that means the handle became read only because we hit our cache limit. So we basically should check if now we can create a cache object (i.e cache cleared out a little)